### PR TITLE
feat: add X12 parsing support

### PR DESCRIPTION
**Pull request by CJHwong**
## Reason

Add the ability to parse X12 EDI strings into structured Python dicts and JSON, complementing the existing rendering (Python -> X12) capability.

## Changes

1. **Parsed data structures** (`pyx12lib/core/parsed.py`): ParsedElement, ParsedCompositeElement, ParsedSegment, ParsedLoop with validation and serialization
2. **Delimiter detection** (`pyx12lib/core/delimiters.py`): Auto-detect element/component/segment delimiters from ISA header
3. **Grammar registry** (`pyx12lib/core/registry.py`): Map segment IDs to grammar definitions with default envelope segments
4. **Parsers** (`pyx12lib/core/parser.py`): SegmentParser for single segments, X12Parser for complete documents
5. **Public API** (`pyx12lib/__init__.py`): `parse_x12()` and `parse_x12_to_json()` convenience functions
6. **Docs** (`README.md`, `CHANGELOG.md`, `setup.py`): Usage examples, changelog, version bump to 0.4

## Test Scope

- 203 tests total, all passing (0.14s)
- Unit tests for each core module: parsed, delimiters, registry, segment parser, x12 parser, composite parser
- EDI 304 validation: 40+ tests with production-grade Shipping Instruction grammars (B2, N1, N7, R4, V1, etc.)
- EDI 315 validation: 25+ tests with Status Detail grammars (B4, R4, DTM, N9, Q2)
- Full-document integration tests including multi-transaction interchanges
- Edge cases: empty input, unknown segments, extra/fewer elements, CRLF separators, performance (2000 segments < 1s)

## Checks

- [x] Unit tests are included
- [x] Integration tests are included
- [x] Documentation updated (README, CHANGELOG)
- [x] All tests passing

### CODE REVIEW

**Review by darwang1 — APPROVED** [review]
From Claude code: One of the cleanest, most well-tested PRs I've reviewed. The code quality is excellent, test coverage is comprehensive, and there are zero backward compatibility concerns.

File: pyx12lib/core/parser.py
Comment (mattwang44) [inline pyx12lib/core/parser.py:1]: nit: I think the API would read better if the parser's configuration (registry) is separated from the input data. Currently `x12_string` is passed to the constructor, making the parser single-use.

Proposed:

```python
parser = X12Parser(registry=registry)
parsed = parser.parse(x12_string)
parsed.to_dict()
parsed.to_json()
```

File: pyx12lib/core/parsed.py [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+
+class ParsedComponent(ParsedElement):
+    """Represents a parsed component within a composite element."""
+    pass
+
+
+class ParsedCompositeElement:
+    """Represents a parsed composite element containing components."""
+
+    def __init__(self, grammar, components):
```
Comment (mattwang44) [inline pyx12lib/core/parsed.py]: nit: type hint

```suggestion
    def __init__(self, grammar: BaseElement, components: Iterable[ParsedComponent]):
```

File: pyx12lib/__init__.py [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+                  segments (ISA/IEA/GS/GE/ST/SE) if not provided.
+
+    Returns:
+        Dict with 'segments' key containing list of parsed segment dicts.
+    """
+    parser = X12Parser(x12_string, registry=registry)
+    return parser.to_dict()
+
+
+def parse_x12_to_json(x12_string, indent=2, registry=None):
```
Comment (mattwang44) [inline pyx12lib/__init__.py]: no indent is more like the default behavior to me (like `json.dumps()`)
```suggestion
def parse_x12_to_json(x12_string, indent=None, registry=None):
```

**Review by mattwang44 — DISMISSED** [review]

File: pyx12lib/core/parsed.py (hunk 1 above)
Comment (CJHwong) [reply pyx12lib/core/parsed.py]: added, thanks!

File: pyx12lib/__init__.py (hunk 2 above)
Comment (CJHwong) [reply pyx12lib/__init__.py]: agreed, revised, thanks!

File: pyx12lib/core/parser.py
Comment (CJHwong) [reply pyx12lib/core/parser.py:1]: I like your style more. refactored, thanks.
